// react-formik/dist/react-formik.esm.js
import React, { useState, useContext, useEffect, createContext } from 'react';
import { useFormikContext, Formik as Formik$1 } from 'formik';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}
function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

var _excluded = ["formikKey", "children"];
var FormikContext = /*#__PURE__*/createContext();
function FormikProvider(_ref) {
  var children = _ref.children;
  var _useState = useState({}),
    states = _useState[0],
    setStates = _useState[1];
  return React.createElement(FormikContext.Provider, {
    value: {
      states: states,
      setStates: setStates
    }
  }, children);
}
function useFormik(props) {
  if (!props) return useFormikContext();
  var formikKey = props.formikKey;
  var context = useContext(FormikContext);
  if (typeof context === "undefined") return undefined;
  var states = context.states,
    setStates = context.setStates;
  var state = states[formikKey] || {};
  return _extends({}, state, {
    setState: function setState(state) {
      setStates(function (prevStates) {
        var _extends2;
        return _extends({}, prevStates, (_extends2 = {}, _extends2[formikKey] = _extends({}, prevStates[formikKey], state), _extends2));
      });
    }
  });
}
function Formik(_ref2) {
  var formikKey = _ref2.formikKey,
    children = _ref2.children,
    props = _objectWithoutPropertiesLoose(_ref2, _excluded);
  var _useFormik = useFormik({
      formikKey: formikKey
    }),
    setState = _useFormik.setState;
  return React.createElement(Formik$1, Object.assign({
    key: formikKey
  }, props), function (state) {
    useEffect(function () {
      setState(state);
    }, [state.dirty, state.errors, state.isSubmitting, state.isValid, state.isValidating, state.status, state.touched, state.values, state.enableReinitialize, state.isInitialValid, state.initialErrors, state.initialStatus, state.initialTouched, state.initialValues]);
    if (typeof children === "function") {
      return children(state);
    }
    return children;
  });
}

export { Formik, FormikProvider, useFormik };
